### ISSUE TITLE
Spotify browsing issues without internet connection

### ISSUE THREAD

**Issue by daniel1v** [hufman/AndroidAutoIdrive#233]
When AndroidAutoIdrive is started without an internet connection on the phone, the Spotify browsing is broken, even for offline content.

First there are no items available at all under "Browse". After refreshing several times, the offline playlists appear directly on the top level of the "Browse" menue. But only the playlists appear, not the other menue items, like "Artists", "Albums" etc. 

Even when the offline playlists appear, most of the times the "Liked Songs" playlist keeps missing, although it is completely offline available. Sometimes it is there and also playable. Unfortunately I was not able to recognize a pattern. 

I cannot remember exactly what worked before, but I am very sure that some of the issues came along with the Spotify Web Integration. When there is an internet connection, everything works fine most of the time. So maybe the case was not considered, that there is no internet connection when starting to drive? For me this is very common, for example in most of the underground parking garages.

Of course this is only a suggestion, there were also several Spotify App updates in the meantime, which could also be the cause of the problems.

Edit: I am using the most recent nightly build.

**Comment by bogdan-calapod** [comment]
That seems to be expected - from my understanding, the browsing is done via the Spotify Web API, which is unavailable while offline. I think @hufman can confirm this.

I think this could be solved by falling back to the old implementation of calling the Spotify SDK when the Web API errors out ?

**Comment by TBGR** [comment]
> That seems to be expected - from my understanding, the browsing is done via the Spotify Web API, which is unavailable while offline. I think @hufman can confirm this.
> 
> I think this could be solved by falling back to the old implementation of calling the Spotify SDK when the Web API errors out ?

For me it has always been like this - no Browse and no Now Playing when offline. If the phone loses data connection after starting Spotify up it keeps the browse and queue functionality until the next time I start up the car.

**Comment by daniel1v** [comment]
> > That seems to be expected - from my understanding, the browsing is done via the Spotify Web API, which is unavailable while offline. I think @hufman can confirm this.
> > I think this could be solved by falling back to the old implementation of calling the Spotify SDK when the Web API errors out ?
> 
> For me it has always been like this - no Browse and no Now Playing when offline. If the phone loses data connection after starting Spotify up it keeps the browse and queue functionality until the next time I start up the car.

I think this depends on your offline content in Spotify. Downloaded playlists used to show up reliably for me, even when the phone was offline, until the Web API was implemented. Since then it stopped working for me. Most of the time I start my car I have no internet connection, so this heavily downgrades my user experience. So if I could 'vote', I would definitely go for a fallback solution.

**Comment by hufman** [comment]
Thank you so much for testing and providing feedback! It is impossible to test all possible scenarios, so I sincerely appreciate your help with this!
Currently, almost everything is still handled through the Spotify SDK access, including Browsing. The Web API is used for rewriting the Liked Songs queue (only after it is selected) to include references to cover art, and for the Search functionality. I suppose there could be the possibility that the code is a little too-tightly coupled, so we'll look into it to be sure.

One possibility that I suspect: When a client app connects through the Spotify SDK, the Spotify app reaches out through the internet to confirm that the user has granted access and hasn't revoked access for this client app. I'm unsure of how long this result is cached, and so it may be the case that the SDK connection is failing to be authenticated at this step.
The Music tab in the phone interface has a checkbox saying "Ability to browse your Spotify library", and this is linked directly to whether the Spotify SDK has connected. In fact, any time you open this tab, it makes an attempt to connect to Spotify to confirm this ability. If the connection was unsuccessful for any reason, it should add a line of text with the error message.
The next time you are in the parking garage without internet access and this misbehavior appears, please open up the Music tab in the phone interface and see if it shows an error message about connecting to the Spotify SDK.

It is definitely possible that Spotify may have changed this behavior, and installing a previous version of Spotify might be a possible test. As another test, I [rebuilt the app](https://androidautoidrive.s3.amazonaws.com/hufman/AndroidAutoIdrive/androidautoidrive-latest-before_spotify_webapi-nomap-sentry-release.apk) from right before merging the SDK access; please try it and see if the behavior changes. Keeping in mind the possible caching behavior of the SDK access above, feel free to try it for several days to see if the same problem shows up.

Let me know what more information you find! Thank you again for your patience in testing!

**Comment by daniel1v** [comment]
So @hufman, I made some tests with your remarks in mind and I think there are at least two different issues. There Was no difference between the most recent nightly build and the Version you posted here before. Both issues can be reproduced without car, by turning on airplane mode and force quitting AndroidAutoIdrive. Then  after opening it manually, I experienced the following issues:


Issue 1:
The first time, I had indeed the issue, that you described, where the Spotify SDK got no access to control Spotify App. So a button appeared in AndroidAutoIdrive, to grant access. I turned off Airplane Mode, made sure, that there is a working WIFI connection and pushed the button. Nothing happened at first. After some waiting time an additional line with an error message appeared, that the access could not be granted (sorry, I cannot remember the exact wording) . I repeatedly pushed the button to grant access, but I could not get it to work, unless I force quitted AndroidAutoIdrive, erased the cache and restarted it with working Internet connection. As you wrote before: This issue seems to only appear, when the access token has expired. For me it happened twice the last two days. But there were also several Re install between, because of my tests.

Issue 2:
The following issue happens every time, when I start AndroidAutoIdrive without internet access. As written before, with Airplane Mode and force quitting you should be able to reproduce it. Basically there is a connection to the Spotify App and also no error message appears. The "browse" tab in the app is empty and also keeps bei g empty without user interaction. After some waiting time and several refreshes by pull down gesture, the offline playlists are appear, but there are not the usual categories like Artists / Albums /Playlists etc.. Also my "Liked Songs" playlist is missing, although it is offline available.

**Comment by hufman** [comment]
Thank you very much for the help! Your very clear reproduction steps helped me see the same problem!

The first issue you saw, which I think is that Spotify didn't bring up the authentication popup after restoring connectivity, could be a delay in Spotify noticing the connectivity. If you have Spotify's Browse tab open, it shows a bar at the bottom saying something like No Internet Connection, and this bar seems to take an extra 30 seconds to notice that the Internet Connection came back. It might have been the case that the Spotify app did not believe it had internet at the time that you tried clicking Authorize, and so it didn't bring the popup up.

I was able to recreate the second issue, both with Airplane Mode and by setting Spotify to Offline Mode. I think the missing Liked Songs is a misbehavior on Spotify's part: Spotify's own Browse tab hid the Liked Songs playlist too when offline, even after enabling the Download toggle. When I flipped the Airplane mode off and swiped down to refresh in AndroidAutoIdrive's Browse tab, after waiting for Spotify to notice the Internet came back, the browse results changed to the proper online entries. I tested in the car too, and toggling Offline Mode changes the car's Browse view similarly.
I even saw the same Browse behavior you reported, where only the offline Playlists are shown at the top level, not the normal Playlist/Artist/Album hierarchy. The initial empty results you saw were probably due to Spotify timing out before realizing the internet is down, before it switched to returning the list of downloaded playlists on the next refresh.

Because the Browse view is showing results at all, this shows that the Spotify SDK is properly connecting to Spotify. I don't believe I have automatic SDK reconnection implemented, so this means that the initial SDK connection when the car app started up was successful, and wasn't failing due to my theory about needing internet to validate the SDK client.

So in summary: I think Spotify stumbles a bit on the transition from online in the house to offline in the garage, showing empty Browse results due to a timeout. After it flips to Offline mode, it shows the list of downloaded playlists, and the missing Liked Songs is due to a bug in Spotify where not even the official Spotify app will show the Liked Songs when offline.

It might be possible (since we've got a pile of hacks to support Spotify's quirks, might as well add another!) to detect when Spotify's top-level Browse is returning a (possibly empty) list of playlists and manually add Liked Songs to this, and it might even accept the request to play that playlist even if it isn't in the browse list. I'll try it out and see what happens!

**Comment by TBGR** [comment]
I don't know if it is relevant, but I noticed that when I have internet connection Spotify tends to load album art a bit slower, sometimes even loading only the upper half. When offline it pops instantly. All my playlists are downloaded.

**Comment by daniel1v** [comment]
@hufman 

so, after several tests, i think a lot of (my) spotify related issues are caused by aai losing the permission to control spotify, because aai was not used for some time. to obtain the persmission back, an internet connection is necessary, as you described before. but also after some time of driving, when internet connection is definitely available, this does not work autotmatically: i have to manually press the "authorize" button on aai or at least restart aai, to get it working.

to circumvent such issues: would it be possible to implement a background service, which automatically renews the spotify permission, so that it is always ensured, that the permission is given? maybe this could be done on a daily base or whenever the phone is connected to wifi and also plugged in.

what do you think about this idea?

**Comment by Kay0002** [comment]
Hi @hufman, this isn't directly related to this issues , but I figured it would be a good to get your feedback before opening an issue.
![IdriveIcons](https://user-images.githubusercontent.com/77034203/114961691-54eadc80-9e37-11eb-8c78-6bc47eb1d0ef.jpg)

I was using 8.5.64.1046 version of Spotify (and have been for several months without issue), however a week or so ago the official Spotify app disappeared from iDrive (I only see the one created by AndriodAutoIdrive app now).   I had a similar issue a month or so ago when the official iHeartRadio app disappeared (similarly I only see the one created by AndriodAutoDrive app now). 

I do still get the iDrive functionality to select from my playlists using the AndriodAutoIdrive icons, but I am wondering why have the official apps suddenly disappeared? 

Apps #3 and #6 in the picture, no longer appear in my IDirve.
![IdriveIcons](https://user-images.githubusercontent.com/77034203/114961723-63d18f00-9e37-11eb-89ae-ba917b38e2c1.jpg)

**Comment by hufman** [comment]
Did you upgrade to MyBMW? The official apps rely on a CarAPI SDK that only works with BMW Connected, and so you'll need to downgrade to BMW Connected 6.4 as opposed to upgrading to MyBMW.

**Comment by Kay0002** [comment]
> Did you upgrade to MyBMW? The official apps rely on a CarAPI SDK that only works with BMW Connected, and so you'll need to downgrade to BMW Connected 6.4 as opposed to upgrading to MyBMW.

Good question. I'm using Connected app v6.5.0.6411, I've been using this version for several months (so no recent change).

**Comment by hufman** [comment]
Very interesting! I'm not sure specifically, I'd suggest trying 6.4 just in case, but otherwise I don't have any other ideas...
The BMW Connected change seemed to happen through a server-side setting, so you might not have noticed any upgrade before it's behavior changed.

**Comment by Kay0002** [comment]
@hufman, as usual you are spot on! Thank you 🙂

I downgraded to Connected 6.4.1.6349 and the official Spotify app (as late as v8.5.66.1002) appears. Oddly, the official iHeartRadio app did not return 🤔

I am really tired of BMWs poor infotainment integration....

**Comment by daniel1v** [comment]
> @hufman, as usual you are spot on! Thank you 🙂
> 
> I downgraded to Connected 6.4.1.6349 and the official Spotify app (as late as v8.5.66.1002) appears. Oddly, the official iHeartRadio app did not return 🤔
> 
> I am really tired of BMWs poor infotainment integration....

@Kay0002 I am very happy to hear, that your problem could be solved. 

For the future I would ask you politely to open a separate issue or use the gitter chat, especially when you know it before, that your issue is not related. It does not make anything better, to use unrelated existing issues. It just messes up the original issue and sends notifications to people who subscribed to the original issue. Also other people with the same issue as yours, are not able to find the solution, because it is hidden inside another issue. 

Please do not understand this as an offense, just a friendly notice for the future.

**Comment by Kay0002** [comment]
@daniel1v that is fair. Where do I find gitter chat? Haven't heard of it and nothing comes up when I search the name.

**Comment by bogdan-calapod** [comment]
@Kay0002 here you go - https://gitter.im/AndroidAutoIdrive/community

**Comment by daniel1v** [comment]
@Kay0002 you can find the gitter chat link in the project description. There is a small grey Button ("badge") on the top, which says "chat on gitter".

https://gitter.im/AndroidAutoIdrive/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge

**Comment by Kay0002** [comment]
@daniel1v @bogdan-calapod, thanks guys! Appreciate it 🙂

**Comment by hufman** [comment]
In recent tests, it seems that Spotify's browse functionality seems to provide more results, and the top-level browse has regular results, so I think this can be closed.
